src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -0,0 +1,216 @@
+use crate::config::{DataType, SinkConfig, SinkContext, SinkDescription};
+use crate::event::MetricValue;
+use crate::http::HttpClient;
+use crate::sinks::gcp::GcpMonitoredResource;
+use crate::sinks::util::http::{BatchedHttpSink, HttpSink};
+use crate::sinks::util::{
+    BatchConfig, BatchSettings, BoxedRawValue, JsonArrayBuffer, TowerRequestConfig,
+};
+use crate::sinks::{Healthcheck, VectorSink};
+use crate::tls::{TlsOptions, TlsSettings};
+use crate::Event;
+use chrono::{DateTime, SecondsFormat, Utc};
+use futures::sink::SinkExt;
+use futures::FutureExt;
+use http::header::AUTHORIZATION;
+use http::{HeaderValue, Uri};
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct StackdriverConfig {
+    pub project_id: String,
+    pub resource: GcpMonitoredResource,
+    pub service_account_file: Option<String>,
+    #[serde(default)]
+    pub request: TowerRequestConfig,
+    #[serde(default)]
+    pub batch: BatchConfig,
+    pub tls: Option<TlsOptions>,
+}
+
+impl_generate_config_from_default!(StackdriverConfig);
+
+inventory::submit! {
+    SinkDescription::new::<StackdriverConfig>("gcp_stackdriver_metrics")
+}
+
+struct HttpEventSink {
+    config: StackdriverConfig,
+    started: DateTime<Utc>,
+    token: gouth::Token,
+}
+
+#[async_trait::async_trait]
+impl HttpSink for HttpEventSink {
+    type Input = serde_json::Value;
+    type Output = Vec<BoxedRawValue>;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        let metric = event.into_metric();
+        let namespace = metric
+            .series
+            .name
+            .namespace
+            .unwrap_or_else(|| "namespace".to_string());
+        let metric_type = format!(
+            "custom.googleapis.com/{}/metrics/{}",
+            namespace, metric.series.name.name
+        );
+        let mut metric_labels = std::collections::HashMap::new();
+
+        for (key, value) in metric.series.tags.unwrap_or_default() {
+            metric_labels.insert(key, value);
+        }
+
+        let metric_kind;
+        let mut interval = std::collections::HashMap::new();
+
+        let end_time = metric.data.timestamp.unwrap_or_else(chrono::Utc::now);
+
+        let point_value = match metric.data.value {
+            MetricValue::Counter { value } => {
+                metric_kind = "CUMULATIVE";
+                interval.insert(
+                    "startTime",
+                    self.started.to_rfc3339_opts(SecondsFormat::Nanos, true),
+                );
+                interval.insert(
+                    "endTime",
+                    end_time.to_rfc3339_opts(SecondsFormat::Nanos, true),
+                );
+                value
+            }
```

Comment:
I'm not a big fan of multiple assignments like this, and would recommend something like the following (though it's not a big deal):
```rust
let (point_value, metric_kind) = match … {
    something => (value, "CUMULATIVE"),
};
```